client/src/hooks/datafiles/mutations/useUpload.ts

```diff
@@ -0,0 +1,118 @@
+import { useSelector, useDispatch, shallowEqual } from 'react-redux';
+import { apiClient } from 'utils/apiClient';
+import Cookies from 'js-cookie';
+import truncateMiddle from 'utils/truncateMiddle';
+import { useMutation } from '@tanstack/react-query';
+
+export async function uploadUtil({
+  api,
+  scheme,
+  system,
+  path,
+  file,
+}: {
+  api: string;
+  scheme: string;
+  system: string;
+  path: string;
+  file: FormData;
+}): Promise<{ file: any; path: string }> {
+  let apiPath = !path || path[0] === '/' ? path : `/${path}`;
+  if (apiPath === '/') {
+    apiPath = '';
+    return { file, path: apiPath };
+  }
+  const formData = new FormData();
+  const fileField = file.get('uploaded_file') as Blob;
+  formData.append('uploaded_file', fileField);
+  let url = `/api/datafiles/${api}/upload/${scheme}/${system}/${apiPath}/`;
+  url.replace(/\/{2,}/g, '/');
+  const response = await apiClient.post(url, formData, {
+    headers: {
+      'X-CSRFToken': Cookies.get('csrftoken') || '',
+    },
+    withCredentials: true,
+  });
+  return response.data;
+}
+
+function useUpload() {
+  const dispatch = useDispatch();
+  const status = useSelector(
+    (state: any) => state.files.operationStatus.upload,
+    shallowEqual
+  );
+
+  const setStatus = (newStatus: any) => {
+    dispatch({
+      type: 'DATA_FILES_SET_OPERATION_STATUS',
+      payload: { status: newStatus, operation: 'upload' },
+    });
+  };
+
+  const { mutate } = useMutation({ mutationFn: uploadUtil });
+
+  const upload = ({
+    system,
+    path,
+    files,
+    reloadCallback,
+  }: {
+    system: string;
+    path: string;
+    files: { data: File; id: string }[];
+    reloadCallback: () => void;
+  }) => {
+    const api = 'tapis';
+    const scheme = 'private';
+
+    files.forEach((fileObj) => {
+      const { data: file, id: index } = fileObj;
+      dispatch({
+        type: 'DATA_FILES_SET_OPERATION_STATUS_BY_KEY',
+        payload: { status: 'UPLOADING', key: index, operation: 'upload' },
+      });
+
+      const formData = new FormData();
+      formData.append('uploaded_file', file);
+
+      mutate(
+        {
+          api,
+          scheme,
+          system,
+          path,
+          file: formData,
+        },
+        {
+          onSuccess: () => {
+            dispatch({
+              type: 'DATA_FILES_SET_OPERATION_STATUS_BY_KEY',
+              payload: { status: 'SUCCESS', key: index, operation: 'upload' },
+            });
+          },
+          onError: (error) => {
+            console.log('The Error is', error);
+            dispatch({
+              type: 'DATA_FILES_SET_OPERATION_STATUS_BY_KEY',
+              payload: { status: 'ERROR', key: index, operation: 'upload' },
+            });
+          },
+        }
+      );
+    });
+
+    dispatch({
```

Comment:
You might want to use `mutateAsync` here so that you can wait for all uploads to complete before dispatching the toast.

Reply:
Changes applied!
https://github.com/TACC/Core-Portal/pull/999/commits/8ffbfd19f401d228bef2893bc1c5f5128ec9bd4a